### PR TITLE
Rearrange Worker model to allow for different kinds of monitoring

### PR DESCRIPTION
**Pull request by alicefuzier**
So that we can add tracing to both pipelines

### CODE REVIEW

**Comment by warrd** [comment]
I'm guessing this means we are planning to update the catalogue apps to latest `messaging` lib? Fingers crossed not too difficult in most of the apps, but think that might be pretty long to do for the `sierra_adapter` (which is currently on very old versions of `messaging` / `storage`).

Something I really want in `messaging` is a base `Stream` trait, so we don't always need to run SQS / SNS in our tests to use `SQSStream`, and use a `MemoryStream` or something instead. I might have a go doing that so we can put that in at the same time.

File: messaging/src/main/scala/uk/ac/wellcome/messaging/worker/steps/MonitoringProcessor.scala [hunk 1]
```diff
@@ -1,28 +1,10 @@
...
-      _: Unit <- metric(result, startTime)
-    } yield Successful[Unit]()
-
-    monitoring recover {
-      case e => MonitoringProcessorFailure[Unit](e)
-    }
-  }
+trait MonitoringProcessor[Message, Context]{
+  def recordStart(message: Message)(implicit ec: ExecutionContext): Future[Context]
+  def recordEnd[Recorded](message: Message, context: Context, result: Result[Recorded])(implicit ec: ExecutionContext): Future[Result[Unit]]
```
Comment (warrd) [inline messaging/src/main/scala/uk/ac/wellcome/messaging/worker/steps/MonitoringProcessor.scala]: Do we run `scalafmt` on this repo? Line length not working out for me :see_no_evil:

**Comment by alicefuzier** [comment]
I might be wrong but I don't think you need a stream if you use the `Worker`.

I agree that updating the Sierra adapter is going to be a pain but we will have to if we want tracing for it

File: messaging/src/main/scala/uk/ac/wellcome/messaging/worker/steps/MonitoringProcessor.scala (hunk 1 above)
Comment (alicefuzier) [reply messaging/src/main/scala/uk/ac/wellcome/messaging/worker/steps/MonitoringProcessor.scala]: Ah I though Travis would that but I guess not..

**Comment by warrd** [comment]
> I might be wrong but I don't think you need a stream if you use the Worker.

Hmmm all our "worker service" classes require either a `SQSStream` or `BigMessageStream` (which in turn uses `SQSStream`). e.g:

* https://github.com/wellcometrust/catalogue/blob/master/pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/services/MergerWorkerService.scala#L17
* https://github.com/wellcometrust/catalogue/blob/master/pipeline/recorder/src/main/scala/uk/ac/wellcome/platform/recorder/services/RecorderWorkerService.scala#L22
* https://github.com/wellcometrust/catalogue/blob/master/mets_adapter/mets_adapter/src/main/scala/uk/ac/wellcome/mets_adapter/services/MetsAdapterWorkerService.scala#L29

etc etc

If we had a base `Stream` trait we could pass in a `MemoryStream` or something instead, in order to test them. Does that make sense, or is there something I'm missing?

**Comment by alicefuzier** [comment]
I think the `Worker` interface abstracts the stream and effectively replaces it. I think our worker services should change to accept a `Worker` which will be a `AlpakkaSQSWorker` in prod code but could be something else in test code

**Comment by alicefuzier** [comment]
If you check the storage service it doesn't use the `SQSStream`. Instead services have an instance of `AlpakkaSQSWorker`  ie [here](https://github.com/wellcomecollection/storage-service/blob/master/ingests/src/main/scala/uk/ac/wellcome/platform/archive/ingests/services/IngestsWorker.scala#L35). The services take care of starting it. It could be injected instead of instantiated inside the service and that would give you the ability to pass whichever `Worker` you want in tests

**Comment by alicefuzier** [comment]
pff Apparently there were a lot of not formatted files, so now the diff has increased significantly. Sorry, everyone

File: messaging/src/test/scala/uk/ac/wellcome/messaging/worker/steps/MonitoringProcessorTest.scala [hunk 2]
```diff
@@ -1,86 +0,0 @@
-package uk.ac.wellcome.messaging.worker.steps
```
Comment (alicefuzier) [inline messaging/src/test/scala/uk/ac/wellcome/messaging/worker/steps/MonitoringProcessorTest.scala]: This has become messaging/src/test/scala/uk/ac/wellcome/messaging/worker/monitoring/metrics/MetricsMonitoringProcessorTest.scala I guess the change is too big for git to notice

File: messaging/src/test/scala/uk/ac/wellcome/messaging/worker/monitoring/metrics/MetricsMonitoringProcessorTest.scala [hunk 3]
```diff
@@ -0,0 +1,100 @@
+package uk.ac.wellcome.messaging.worker.monitoring.metrics
```
Comment (alicefuzier) [inline messaging/src/test/scala/uk/ac/wellcome/messaging/worker/monitoring/metrics/MetricsMonitoringProcessorTest.scala:1]: This used to be MonitoringProcessorTest.scala but the change is too big for git to notice

**Comment by warrd** [comment]
Ah gotcha about the  `Worker`. Will our current code have to be updated to use `Worker` instead of the old stream interface for this monitoring stuff to work?

**Comment by alicefuzier** [comment]
Yep, but this way tracing becomes usable in the storage service too and we achieve a consistent approach in all our services (excluding the sierra adapter which becomes even more difficult to update)

File: messaging/src/main/scala/uk/ac/wellcome/messaging/worker/Worker.scala [hunk 4]
```diff
@@ -1,40 +1,40 @@
...
-
+  private def work(message: Message)(
+    implicit monitoringProcessor: MonitoringProcessor[Message,
+                                                      MonitoringContext])
+    : Future[Completion] = {
     for {
+      context <- monitoringProcessor.recordStart(message)
       summary <- process(message)
-      monitor <- record(startTime, summary)
+      monitor <- monitoringProcessor.recordEnd(message, context, summary)
```
Comment (warrd) [inline messaging/src/main/scala/uk/ac/wellcome/messaging/worker/Worker.scala:37]: It seems to me like if an exception is raised inside the processing of the message, we won't `recordEnd`. Is that correct, and is that the desired behaviour?

Comment (alicefuzier) [reply messaging/src/main/scala/uk/ac/wellcome/messaging/worker/Worker.scala:37]: No: `process` returns a `Future[Result[Summary]]` where `Result` encapsulates whether the `Future` succeeded or failed. Even if there is an exception inside `process` it should be caught and encapsulated

Comment (warrd) [reply messaging/src/main/scala/uk/ac/wellcome/messaging/worker/Worker.scala:37]: Oh yep, I'm guessing you mean [here](https://github.com/wellcomecollection/scala-messaging/blob/14b5fecf362220460d8ec2837ce29f0e37eed773/messaging/src/main/scala/uk/ac/wellcome/messaging/worker/steps/MessageProcessor.scala#L22)

**Review by warrd — APPROVED** [review]
:tada:

**Pull request merged**